### PR TITLE
GEMGeometryBuilderDB-fix

### PR DESCRIPTION
**Pull request by jshlee**
PR #19707 reverted changes made by @calabria.
This PR fixes the GEMGeometryBuilderFromCondDB.

@ianna @kpedro88

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 1]
```diff
@@ -76,65 +80,108 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
   }
-  /*
+  
   // Create the GEMChambers and store them on the Geometry 
 
-  for( std::map<GEMDetId, std::list<GEMEtaPartition *> >::iterator ich=chids.begin();
-       ich != chids.end(); ich++){
-    GEMDetId chid = ich->first;
-    std::list<GEMEtaPartition * > gepls = ich->second;
+  for( auto& ich : m_chids ) {
```
Comment (ianna) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @jshlee - const auto&?

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 2]
```diff
@@ -76,65 +80,108 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
 
     // compute the overall boundplane. At the moment we use just the last
     // surface
-    BoundPlane* bp=0;
-    for(std::list<GEMEtaPartition *>::iterator gepl=gepls.begin();
-    gepl!=gepls.end(); gepl++){
-    const BoundPlane& bps = (*gepl)->surface();
-      bp = const_cast<BoundPlane *>(&bps);
+    BoundPlane* bp = 0;
+    for( auto& gep : gepList ) {
```
Comment (ianna) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @jshlee - const auto&?

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 3]
```diff
@@ -76,65 +80,108 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
+    std::list<GEMEtaPartition * > gepList = ich.second;
 
     // compute the overall boundplane. At the moment we use just the last
     // surface
-    BoundPlane* bp=0;
-    for(std::list<GEMEtaPartition *>::iterator gepl=gepls.begin();
-    gepl!=gepls.end(); gepl++){
-    const BoundPlane& bps = (*gepl)->surface();
-      bp = const_cast<BoundPlane *>(&bps);
+    BoundPlane* bp = 0;
```
Comment (ianna) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: nullptr?

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 4]
```diff
@@ -76,65 +80,108 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
+    GEMChamber* ch = new GEMChamber( chid, surf ); 
+    LogDebug("GEMGeometryBuilder") << "GEM Chamber created with id = " << chid
+				   << " = " << chid.rawId() << " and added to the GEMGeometry" << std::endl;
+    LogDebug("GEMGeometryBuilder") << "GEM Chamber has following eta partitions associated: " << std::endl;
+
     // Add the etaps to rhe chamber
-    for(std::list<GEMEtaPartition *>::iterator gepl=gepls.begin();
-    gepl!=gepls.end(); gepl++){
-      ch->add(*gepl);
+    for( auto& gep : gepList ) {
```
Comment (ianna) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: const auto&?

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 5]
```diff
@@ -76,65 +80,108 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
+      GEMDetId schid( chid.region(), chid.ring(), chid.station(), 0, chid.chamber(), 0 );
+      GEMSuperChamber* sch = new GEMSuperChamber( schid, surf );
+      LogDebug("GEMGeometryBuilder") << "GEM SuperChamber created with id = " << schid
+				     << " and added to the GEMGeometry" << std::endl;
+      superChambers.emplace_back( sch );
+    }
+    
     // Create the chamber 
-    GEMChamber* ch = new GEMChamber (chid, surf); 
+    GEMChamber* ch = new GEMChamber( chid, surf );
```
Comment (ianna) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @jshlee - please, consider using unique or shared pointers

**Review by ianna — CHANGES_REQUESTED** [review]

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc (hunk 5 above)
Comment (jshlee) [reply Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: for all new objects or just this one?

Comment (ianna) [reply Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: for all, please

**Comment by jshlee** [comment]
@ianna - I changed to shared_ptr in the similar fashion that was implemented for the other detectors  as well as your other suggestions.

**Comment by kpedro88** [comment]
@jshlee it looks like you missed some instances in GEMGeometryBuilderFromCondDB.cc where `shared_ptr` should be used

**Comment by jshlee** [comment]
@kpedro88 where at??

**Comment by kpedro88** [comment]
look where @ianna commented... anything allocated with `new` should be a smart pointer instead, to reduce the chance of memory leaks.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc (hunk 5 above)
Comment (kpedro88) [reply Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @ianna did you decide this comment can be ignored?

Comment (ianna) [reply Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @kpedro88 - I'm going to fix all geometry/DD code to use smart pointers, so yes, it's ok to leave some here :-)

**Comment by kpedro88** [comment]
+1

File: Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc [hunk 6]
```diff
@@ -76,65 +79,106 @@ GEMGeometry* GEMGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
...
+	  theGeometry->add( superChamber );
+	  LogDebug("GEMGeometryBuilderFromDDD") << "Adding super chamber " << detId << " to ring: " 
+						<< "re " << re << " st " << st << " ri " << ri << std::endl;
+	}
+	LogDebug("GEMGeometryBuilderFromDDD") << "Adding ring " <<  ri << " to station " << "re " << re << " st " << st << std::endl;
+	station->add( const_cast<GEMRing*>( ring ));
+	theGeometry->add( const_cast<GEMRing*>( ring ));
+      }
+      LogDebug("GEMGeometryBuilderFromDDD") << "Adding station " << st << " to region " << re << std::endl;
+      region->add(const_cast<GEMStation*>(station));
```
Comment (davidlange6) [inline Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: hi @jshlee @ianna - didn't we already review/discuss the same stanza of code with the same const_cast problems once before?

Comment (ianna) [reply Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc]: @davidlange6 - yes, we did and it was fixed here:
https://github.com/jshlee/cmssw/commit/d64f6e0e9e1c8366d11b9513dbb59fdd37dd9304#diff-991663c85a84d7e4630761ba3d046958L163

**Comment by davidlange6** [comment]
so could we not reuse the existing code rather than duplicating it?

**Comment by ianna** [comment]
Sorry, I was not clear. I was talking about line 160 of this PR which was fixed after the discussion a while ago:
```
-	  GEMSuperChamber* superChamber = const_cast<GEMSuperChamber*>( superChambers.at(sch));
+	  GEMSuperChamber* superChamber = superChambers[sch];
```
We did not discuss the rest of const_casts.

**Comment by jshlee** [comment]
now the only const_cast are in https://github.com/cms-sw/cmssw/pull/20260/files#diff-991663c85a84d7e4630761ba3d046958R166

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
@jshlee please rebase to deal with merge conflict
http://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by ianna** [comment]
@jshlee - please rebase

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@smuzaffar is this something broken/out of sync in the IB?

**Comment by smuzaffar** [comment]
@kpedro88 , yes. This PR testing also picked up changed merged in master/93X branch via #20237 but these also needed an external change https://github.com/cms-sw/cmsdist/pull/3374 which is not part of 09-05-2300 IB. I would suggest to wait for today's 11h00 IB and then restart the tests.

**Comment by kpedro88** [comment]
@smuzaffar okay. How do I tell when the new IB is available to be used in PR tests?

**Comment by kpedro88** [comment]
@jshlee in the meantime, a question: is this PR important to include in the HGCal TDR production? If so, I think it will need to be backported to 93X (now that 94X is open).

**Comment by jshlee** [comment]
@kpedro88 - yes this would be needed.

**Comment by kpedro88** [comment]
@jshlee assuming there are no more updates to this branch and it can be merged in master ASAP, then please submit the backport (or I can do it) reusing the same branch:
https://github.com/cms-sw/cmssw/compare/CMSSW_9_3_X...jshlee:GEMGeometryBuilderDB-fix

**Comment by jshlee** [comment]
yep that sounds good.

**Comment by ianna** [comment]
please test

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
@ianna please sign if acceptable

**Comment by ianna** [comment]
+1

**Comment by arunhep** [comment]
@jshlee do you have a backport to 92X ? newly proposed GEM geometry is already deployed in data-taking. So, if you want to have the correct geometry being used in data-taking then this PR should be backported to 92X and should be part of next 92X release.

**Comment by jshlee** [comment]
@kpedro88 @ianna - This should be in 92X too. What do you guys recommend?

**Comment by kpedro88** [comment]
@jshlee you'll have to make a separate branch to backport to 92X. Instructions for that are here: http://cms-sw.github.io/tutorial-resolve-conflicts.html#backporting-a-pr

**Comment by ianna** [comment]
@jshlee - yes, it should be backported to 92x

**Comment by kpedro88** [comment]
@davidlange6 this is a must-have for 9_3_0. We cannot have a geometry regression in a production release because a 3-week-old PR was ignored.

**Comment by davidlange6** [comment]
hi @jshlee - the other const_casts need to be removed (as I commented in the first PR for this code..) - since all the pointers are created in the same function, this should just be a rearrangement - perhaps even the data structure can be built from bottom up as to simply the loop structure?

**Comment by davidlange6** [comment]
-1
pending suggested refactoring

**Comment by ianna** [comment]
@davidlange6 - IMHO, the correct way to remove the const castness is to change the API of GEMSuperChamber:
```
diff --git a/Geometry/GEMGeometry/interface/GEMSuperChamber.h b/Geometry/GEMGeometry/interface/GEMSuperChamber.h
index 8186a97d7a9..d0c3390aea8 100644
--- a/Geometry/GEMGeometry/interface/GEMSuperChamber.h
+++ b/Geometry/GEMGeometry/interface/GEMSuperChamber.h
@@ -38,7 +38,7 @@ class GEMSuperChamber : public GeomDet
   bool operator==(const GEMSuperChamber& sch) const;
 
   /// Add chamber to the super chamber which takes ownership

   /// Add chamber to the super chamber which takes ownership
-  void add(GEMChamber* ch);
+  void add(const GEMChamber* ch);

```
```
diff --git a/Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc b/Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc
index b03800b1507..9ab98793698 100644
--- a/Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc
+++ b/Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc
@@ -161,12 +161,8 @@ GEMGeometryBuilderFromCondDB::build(const std::shared_ptr<GEMGeometry>& theGeome
          const GEMDetId detId( superChamber->id());
          if (detId.region() != re || detId.station() != st || detId.ring() != ri) continue;
          
-         auto ch1 = theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 1, detId.chamber(), 0 ));
-         auto ch2 = theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 2, detId.chamber(), 0 ));
-         superChamber->add( const_cast<GEMChamber*>( ch1 ));
-         superChamber->add( const_cast<GEMChamber*>( ch2 ));
-         // superChamber->add( ch1 );
-         // superChamber->add( ch2 );
+         superChamber->add( theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 1, detId.chamber(), 0 )));
+         superChamber->add( theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 2, detId.chamber(), 0 )));
          ring->add( superChamber );
          theGeometry->add( superChamber );

```

However, this change will bring in another class GEMGeometryBuilderFromDDD.cc which needs to be cleaned from const_casts as well. This will require changes in GEMGeometry API.

**Comment by ianna** [comment]
@davidlange6 - would it be acceptable to merge this PR and move further discussions to #20562?

**Comment by jshlee** [comment]
I'm happy to do either.

**Comment by ianna** [comment]
@davidlange6 ?

**Comment by davidlange6** [comment]
i'd prefer just to work with #20562 - which looks good to me.

**Pull request closed without merging**